### ISSUE TITLE
[Bug]: Gemma4 Unified image requests produce all-NaN logits after BF16-to-FP16 fallback

### ISSUE THREAD

**Issue by TooYoungTooSimp** [vllm-project/vllm#48231]
### Your current environment

<details>
<summary>The output of <code>python collect_env.py</code></summary>

```text
Collecting environment information...
uv is set
==============================
        System Info
==============================
OS                           : Ubuntu 24.04.4 LTS (x86_64)
GCC version                  : (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version                : Could not collect
CMake version                : version 3.28.3
Libc version                 : glibc-2.39

==============================
       PyTorch Info
==============================
PyTorch version              : 2.11.0+cu130
Is debug build               : False
CUDA used to build PyTorch   : 13.0
ROCM used to build PyTorch   : N/A
XPU used to build PyTorch    : N/A

==============================
      Python Environment
==============================
Python version               : 3.13.9 | packaged by Anaconda, Inc. | (main, Oct 21 2025, 19:16:10) [GCC 11.2.0] (64-bit runtime)
Python platform              : Linux-6.17.0-35-generic-x86_64-with-glibc2.39
    
==============================
       CUDA / GPU Info
==============================
Is CUDA available            : True
CUDA runtime version         : Could not collect
CUDA_MODULE_LOADING set to   : 
GPU models and configuration : GPU 0: NVIDIA GeForce RTX 2080 Ti
Nvidia driver version        : Could not collect
cuDNN version                : Could not collect
HIP runtime version          : N/A
MIOpen runtime version       : N/A
Is XNNPACK available         : True

==============================
          CPU Info
==============================
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           48 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  16
On-line CPU(s) list:                     0-15
Vendor ID:                               AuthenticAMD
Model name:                              AMD Opteron(tm) Processor 4386
CPU family:                              21
Model:                                   2
Thread(s) per core:                      1
Core(s) per socket:                      8
Socket(s):                               2
Stepping:                                0
Frequency boost:                         enabled
CPU(s) scaling MHz:                      36%
CPU max MHz:                             3100.0000
CPU min MHz:                             1400.0000
BogoMIPS:                                6200.50
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 popcnt aes xsave avx f16c lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs xop skinit wdt fma4 tce nodeid_msr tbm topoext perfctr_core perfctr_nb cpb hw_pstate ssbd ibpb vmmcall bmi1 arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold
Virtualization:                          AMD-V
L1d cache:                               256 KiB (16 instances)
L1i cache:                               512 KiB (8 instances)
L2 cache:                                16 MiB (8 instances)
L3 cache:                                12 MiB (2 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-7
NUMA node1 CPU(s):                       8-15
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Old microcode:             Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Mitigation; untrained return thunk; SMT disabled
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Retpolines; IBPB conditional; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Not affected

==============================
Versions of relevant libraries
==============================
[pip3] flashinfer-python==0.6.12
[pip3] numpy==2.4.6
[pip3] nvidia-cublas==13.1.0.3
[pip3] nvidia-cuda-cccl==13.3.3.4.1
[pip3] nvidia-cuda-crt==13.3.73
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvcc==13.2.78
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cuda-tileiras==13.2.78
[pip3] nvidia-cudnn-cu13==9.19.0.56
[pip3] nvidia-cudnn-frontend==1.25.0
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-cufile==1.15.1.6
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.0
[pip3] nvidia-cutlass-dsl==4.5.2
[pip3] nvidia-cutlass-dsl-libs-base==4.5.2
[pip3] nvidia-cutlass-dsl-libs-cu13==4.5.2
[pip3] nvidia-ml-py==13.610.43
[pip3] nvidia-nccl-cu13==2.28.9
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvshmem-cu13==3.4.5
[pip3] nvidia-nvtx==13.0.85
[pip3] nvidia-nvvm==13.2.78
[pip3] pyzmq==27.1.0
[pip3] tokenspeed-triton==3.7.10.post20260531
[pip3] torch==2.11.0
[pip3] torch-c-dlpack-ext==0.1.5
[pip3] torchaudio==2.11.0
[pip3] torchvision==0.26.0
[pip3] transformers==5.12.1
[pip3] triton==3.6.0
[conda] pyzmq                      27.1.0           py313hcf8288c_1     defaults

==============================
         vLLM Info
==============================
ROCM Version                 : Could not collect
vLLM Version                 : 0.24.0
vLLM Build Flags:
  CUDA Archs: Not Set; ROCm: Disabled; XPU: Disabled
GPU Topology:
  	[4mGPU0	CPU Affinity	NUMA Affinity	GPU NUMA ID[0m
GPU0	 X 	0-7	0		N/A

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

==============================
     Environment Variables
==============================
PYTORCH_NVML_BASED_CUDA_CHECK=1
TORCHINDUCTOR_COMPILE_THREADS=1
TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_ubuntu


```

</details>

### 🐛 Describe the bug

Gemma4 Unified image requests generate no assistant content when a BF16 checkpoint is served on a platform that does not support BF16 and vLLM automatically falls back to FP16.

The API reports normal generation throughput and increasing KV-cache usage. With a completion limit, it returns `content=null` and `finish_reason="length"`. Without a completion limit, generation continues indefinitely. Enabling or disabling thinking does not change the behavior.

The server reports the expected dtype fallback:

```text
Your device 'NVIDIA GeForce RTX 2080 Ti' (with compute capability 7.5)
doesn't support torch.bfloat16. Falling back to torch.float16 for compatibility.
Casting torch.bfloat16 to torch.float16.
```

### Reproduction

Start the server without speculative decoding:

```bash
vllm serve /path/to/gemma-4-12B-it-qat-w4a16-ct \
  --port 14514 \
  --max-num-seqs 128 \
  --default-chat-template-kwargs '{"enable_thinking": false}'
```

Send a self-contained request using a generated 512x512 RGB PNG:

```python
import base64
import io

from openai import OpenAI
from PIL import Image

buffer = io.BytesIO()
Image.new("RGB", (512, 512), "red").save(buffer, format="PNG", optimize=True)
image_url = "data:image/png;base64," + base64.b64encode(
    buffer.getvalue()
).decode()

client = OpenAI(base_url="http://127.0.0.1:14514/v1", api_key="dummy")
response = client.chat.completions.create(
    model="gemma-4-12B-it-qat-w4a16-ct",
    messages=[{
        "role": "user",
        "content": [
            {"type": "image_url", "image_url": {"url": image_url}},
            {"type": "text", "text": "What color is in this image?"},
        ],
    }],
    max_completion_tokens=1024,
    extra_body={"return_token_ids": True},
)

print(response.choices[0].message.content)
print(response.choices[0].finish_reason)
print(response.choices[0].token_ids[:16])
```

Observed output:

```text
None
length
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
```

The full completion contains 1024 copies of token ID 0. The issue also reproduces with 1920x1080 images, so it is not specific to small images or the Transformers 1x1 channel-dimension warning.

The prompt is formed correctly: returned prompt token IDs contain the image boundaries and 280 expanded image placeholders:

```text
... 255999, 258880, 258880, ... 258880, 258882 ...
```

`bad_words=["<pad>"]`, `logit_bias={"0": -100}`, thinking on/off, streaming/non-streaming, and disabling FlashInfer sampling do not fix the problem.

### Debugging and root cause

Sampler instrumentation immediately after converting logits to FP32:

```python
raw_nan_counts = torch.isnan(logits).sum(dim=-1)
raw_finite_counts = torch.isfinite(logits).sum(dim=-1)
```

Output:

```text
shape=(1, 262144)
raw_nan_counts=[262135]
raw_finite_counts=[0]
```

LM-head instrumentation shows that its weights are finite, but the hidden state entering `compute_logits` is already entirely NaN:

```python
hidden_nan = torch.isnan(hidden_states).sum()
hidden_finite = torch.isfinite(hidden_states).sum()
logits = self.logits_processor(self.lm_head, hidden_states)
```

```text
hidden={'shape': (1, 3840), 'nan': 3840, 'finite': 0}
lm_head_weight={'shape': (262144, 3840), 'sample_nan': 0, 'sample_finite': 64}
logits={'shape': (1, 262144), 'nan': 262144, 'finite': 0}
```

The multimodal path confirms that NaNs exist before the language model:

```text
inputs_embeds={'shape': (278, 3840), 'nan': 983040, 'finite': 84480}
hidden={'shape': (278, 3840), 'nan': 1067520, 'finite': 0}
```

The vision embedder was instrumented at every stage:

```python
cast_pixels = pixel_values.to(self.pos_embedding.dtype)
after_ln1 = self.patch_ln1(cast_pixels)
after_dense, _ = self.patch_dense(after_ln1)
after_ln2 = self.patch_ln2(after_dense)
pos_embs = self._factorized_posemb(pixel_position_ids)
hidden_states = self.pos_norm(after_ln2 + pos_embs)
```

The first non-finite values appear in `patch_dense`:

```text
pixels: dtype=float16, nan=0, inf=0, bad_rows=0
ln1:    dtype=float16, nan=0, inf=0, bad_rows=0
dense:  dtype=float16, nan=0, inf=256, bad_rows=256
ln2:    dtype=float16, nan=983040, inf=0, bad_rows=256
posemb: dtype=float16, nan=0, inf=0, bad_rows=0
```

All 256 `Inf` values occur in output channel 1215:

```text
dense_inf_locations=[[0, 1215], [1, 1215], ... [255, 1215]]
dense_finite_absmax=49248.0
```

The checkpoint tensors are finite but large:

```text
vision_embedder.patch_dense.weight: bf16, nan=0, inf=0, min=-130, max=147
vision_embedder.patch_dense.bias:   bf16, nan=0, inf=0, min=-127, max=126
```

The 6912-dimensional affine projection overflows FP16's maximum finite value of 65504 before LayerNorm can normalize it. The following LayerNorm expands the single `Inf` in each valid row into an all-NaN row, which then propagates through the language model and logits. Sampling token ID 0 is a downstream symptom, not the cause.

### Expected behavior

Image embeddings, language-model hidden states, and logits should remain finite when vLLM serves a Gemma4 Unified checkpoint on a platform where the checkpoint dtype requires fallback.

### Validated fix

When the vision config dtype is not included in `current_platform.supported_dtypes`, load and evaluate `vision_embedder.patch_dense` and its following `patch_ln2` in FP32. Cast the normalized result back to the model dtype before positional embeddings and the remaining multimodal pipeline.

After this change, the same request returns normal assistant content:

```text
vision embedded: 1,075,200 / 1,075,200 finite
valid features:    983,040 /   983,040 finite
LM hidden:       1,067,520 / 1,067,520 finite
decode logits:     262,144 /   262,144 finite
NaN values: 0
```

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.
